packages/react-components/react-motions-preview/src/factories/createTransition.ts

```diff
@@ -0,0 +1,110 @@
+import { useEventCallback, useIsomorphicLayoutEffect, useMergedRefs } from '@fluentui/react-utilities';
+import * as React from 'react';
+
+import { useIsReducedMotion } from '../hooks/useIsReducedMotion';
+import type { MotionTransition } from '../types';
+
+type TransitionProps = {
+  children: React.ReactElement;
+
+  appear?: boolean;
+  visible?: boolean;
+
+  unmountOnExit?: boolean;
+};
+
+// TODO: use Transition types
+export function createTransition(transition: MotionTransition) {
+  const Transition: React.FC<TransitionProps> = props => {
+    const { appear, children, visible, unmountOnExit } = props;
+
+    const child = React.Children.only(children) as React.ReactElement & { ref: React.Ref<HTMLElement> };
+
+    const elementRef = React.useRef<HTMLElement>();
+    const ref = useMergedRefs(elementRef, child.ref);
+
+    const [mounted, setMounted] = React.useState(() => (unmountOnExit ? visible : true));
+
+    const isFirstMount = React.useRef<boolean>(true);
+    const isReducedMotion = useIsReducedMotion();
+
+    const onExitFinish = useEventCallback(() => {
+      if (unmountOnExit) {
+        setMounted(false);
+      }
+    });
+
+    useIsomorphicLayoutEffect(() => {
+      if (visible) {
+        setMounted(true);
+        return;
+      }
+
+      if (elementRef.current) {
+        const animation = elementRef.current.animate(transition.exit.keyframes, {
+          fill: 'forwards',
+
+          ...transition.exit.options,
+          ...(isReducedMotion() && { duration: 1 }),
+        });
+
+        if (isFirstMount.current) {
+          animation.finish();
```

Comment:
Why `finish` and not `cancel` here?

Reply:
`.finish()` just finishes animation, `.cancel()` - just cancels. We need to have animation finished to apply proper styles. For example, when an element is not visible i.e. exited stated, we need to have `opacity: 0`. I added a comment with an explanation to code.

I need to experiment there a bit more, may be I can find a different option there.